uplink/config_test.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"storj.io/storj/pkg/storj"
+)
+
+func TestEncryptionConfig_Key(t *testing.T) {
+	saveKey := func(key []byte) (filepath string, removeFile func()) {
+		f, err := ioutil.TempFile("", "storj-test-uplink-keyfilepath-*")
+		require.NoError(t, err)
+		defer func() { _ = f.Close() }()
+
+		_, err = f.Write(key)
+		require.NoError(t, err)
+
+		err = f.Chmod(os.FileMode(0400))
+		require.NoError(t, err)
+
+		return f.Name(), func() { _ = os.Remove(f.Name()) }
+	}
+
+	t.Run("ok: file with key length less or equal than max size", func(t *testing.T) {
+		someKey := make([]byte, rand.Intn(20)+1)
+		_, err := rand.Read(someKey)
+		require.NoError(t, err)
+		fpath, cleanup := saveKey(someKey)
+		defer cleanup()
+
+		var expKey storj.Key
+		copy(expKey[:], someKey)
+
+		encCfg := &EncryptionConfig{
+			KeyFilepath: fpath,
+		}
+		key, err := encCfg.Key()
+		require.NoError(t, err)
+
+		assert.Equal(t, expKey[:], key[:])
+	})
+
+	t.Run("ok: file with key length greater than max size", func(t *testing.T) {
+		expKey := make([]byte, rand.Intn(10)+1+storj.KeySize)
+		_, err := rand.Read(expKey)
+		require.NoError(t, err)
+		fpath, cleanup := saveKey(expKey)
+		defer cleanup()
+
+		encCfg := &EncryptionConfig{
+			KeyFilepath: fpath,
+		}
+		key, err := encCfg.Key()
+		require.NoError(t, err)
+
+		assert.Equal(t, expKey[:storj.KeySize], key[:])
+	})
+
+	t.Run("ok: empty file path", func(t *testing.T) {
+		fpath, cleanup := saveKey([]byte{})
+		defer cleanup()
+
+		encCfg := &EncryptionConfig{
+			KeyFilepath: fpath,
+		}
+		key, err := encCfg.Key()
+		require.NoError(t, err)
+		assert.Equal(t, key, storj.Key{})
+	})
+
+	t.Run("error: file not found", func(t *testing.T) {
+		// Create a temp file and delete it, to get a filepath which doesn't exist.
+		f, err := ioutil.TempFile("", "storj-test-uplink-keyfilepath-*")
+		require.NoError(t, err)
+		err = f.Close()
+		require.NoError(t, err)
+		err = os.Remove(f.Name())
+		require.NoError(t, err)
+
+		encCfg := &EncryptionConfig{
+			KeyFilepath: f.Name(),
+		}
+		_, err = encCfg.Key()
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), fmt.Sprintf("not found key file %q", f.Name()))
+		assert.True(t, Error.Has(err), "err is not of %q class", Error)
+	})
+
+	t.Run("error: permissions are too open", func(t *testing.T) {
+		// Create a key file and change its permission for not being able to read it
+		f, err := ioutil.TempFile("", "storj-test-uplink-keyfilepath-*")
+		require.NoError(t, err)
+		defer func() {
+			_ = f.Close()
+			_ = os.Remove(f.Name())
```

Comment:
errors shouldn't be ignored

Reply:
Same reason than before, it's a test, these errors don't provide anything useful to report to the test if they are returned nor stop the test to run

Reply:
its more a convention that we are handling all errors, even in tests to be sure we are not leaving junk or unclosed things

Reply:
I understand what you mean, because it's what I would like but in this cases my questions are:

1. Would you like that the CI stop because one temporary file created by the test isn't deleted or closing its file descriptor has failed?
2. If the test in the CI fails because the temporary file, created by the test, has not been deleted, do you think that we are going to take care to go and delete such temporary file manually? If I consider that the CI spins up a new container on each run (I'm not sure about this, it's a guess) and the disk is ephemeral, then I'm sure that no need, as it won't remain.

If the convention is, we don't want to think in if it makes sense to skip the error or not, then I'm fine with that, only note that if I wrote `_` is because I was aware of the errors that those functions can return otherwise I would have written `f.Close()` and `os.Remove(..)`.

Reply:
Yes, we want the CI tests to fail if any of the hypothesis you've described happen. Every time an unexpected error happens we need to know about it and investigate it. This is why we check all errors.

Reply:
The important thing on those hypothesis are **"one temporary file created by the test"** if that wouldn't come from the test there wouldn't be any doubt.

However, the true is that if the test has a bug in terms of calling those functions at the wrong moment, then it could be an error but ignoring them, we will never know about such bug, obviously at the expense of the CI stoping because of an unlikely error returned by the OS, which mostly could warn about a problem with the CI machine although we would probably have other problems at that same moment too.

Adding those assertions, considering that one line (e.g. `require.NoError(t, f.Close())`) is readable/clear/understandable.

Reply:
`os.Remove` gives very useful information on Windows. When it fails then there's something is still using the file, which means we have a leak in our system.

Or in other words, we have fixed production leaks by starting to check these errors.

Reply:
> Or in other words, we have fixed production leaks by starting to check these errors.

It's a test, so this error wont' happen in production, otherwise I won't have arguments.

Anyway, after know that we are OK on stopping the CI for these matters, then it looks good to assert them.

Reply:
I mean a test detecting a leak that was in production. It's not about leaking the specific file, but an open file descriptor to that file in an unrelated place. For example, let's say you forget to close the file in `ReadKey` then `os.Remove` will fail on Windows.

Reply:
That's a good example of the bug case that I mentioned above.